### PR TITLE
MGMT-13001: Adding cost by job type

### PR DESCRIPTION
**Pull request by danmanor**
Adding cost by job type to `equinix` section of the report.

image

### CODE REVIEW

File: src/prowjobsscraper/indices/usages_schema.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{
+    "settings": {
+      "index": {
+        "number_of_shards": "1",
+        "number_of_replicas": "0",
```
Comment (rccrdpccl) [inline src/prowjobsscraper/indices/usages_schema.json:5]: just out of curiosity - why 0 replicas?

Comment (danmanor) [reply src/prowjobsscraper/indices/usages_schema.json:5]: I followed the other sachems, I am not sure. @adriengentil ?

Comment (adriengentil) [reply src/prowjobsscraper/indices/usages_schema.json:5]: I guess it just happened to be 0 :D
@rccrdpccl how many replicas would you recommend? They are all small indices.

Comment (rccrdpccl) [reply src/prowjobsscraper/indices/usages_schema.json:5]: 0 replicas is vulnerable for data loss if something goes wrong in the node where the shard is assigned.
As it's small and we have no concerns of filling up the disk, I'd recommend 2 replicas (plus primary).
With the caveat that in integration we only have one node so it would need to be 0 there, otherwise it be marked as "yellow".

In any case, it's not important as we also have backups :)

File: src/jobsautoreport/report.py [hunk 2]
```diff
@@ -12,7 +12,22 @@
...
+class Metric(BaseModel):
+    def is_zero(self) -> bool:
+        for cost in self.dict().values():
+            if cost > 0:
+                return False
+        return True
+
+
+class JobTypeMetrics(Metric):
+    presubmit: float = 0
```
Comment (adriengentil) [inline src/jobsautoreport/report.py:24]: same comment about type annotation as in the other PR

File: src/jobsautoreport/report.py [hunk 3]
```diff
@@ -271,6 +281,30 @@ def _count_cost_by_machine_type(
...
+            periodic=cls._count_cost_by_job_type(usages, jobs, JobType.PERIODIC),
+            postsubmit=cls._count_cost_by_job_type(usages, jobs, JobType.POSTSUBMIT),
+            batch=cls._count_cost_by_job_type(usages, jobs, JobType.BATCH),
+        )
+
+    @staticmethod
+    def _count_cost_by_job_type(
+        usages: list[EquinixUsageEvent], jobs: list[JobDetails], job_type: JobType
+    ) -> float:
+        jobs_build_id_to_type = {job.build_id: job.type for job in jobs}
```
Comment (adriengentil) [inline src/jobsautoreport/report.py:299]: just get a set where all build_ids are from the selected type
then later in `if usage.job.build_id in jobs_build_id_matching_type`

File: src/prowjobsscraper/scraper.py [hunk 4]
```diff
@@ -71,10 +71,17 @@ def _should_index_usage(
         known_usages_identifiers: set[equinix_usages.EquinixUsageIdentifier],
         usages: list[equinix_usages.EquinixUsage],
     ) -> bool:
```
Comment (danmanor) [inline src/prowjobsscraper/scraper.py:73]: This part is not related to the `cost by job type`, I added it to the PR

File: src/jobsautoreport/models.py [hunk 5]
```diff
@@ -24,8 +24,40 @@ class ReportInterval(Enum):
...
+    successes: int
+    failures: int
+
+    @property
+    def total(self) -> int:
+        return self.successes + self.failures
+
+    @property
+    def failure_rate(self) -> float:
+        return 0 if self.total == 0 else (self.failures / self.total) * 100
```
Comment (osherdp) [inline src/jobsautoreport/models.py:59]: mathematically it should not be defined (like 'N/A'), because 0 is a valid failure rate only if you have any kind of attempts 
```suggestion
        return None if self.total == 0 else (self.failures / self.total) * 100
```

**Review by osherdp — CHANGES_REQUESTED** [review]

File: src/jobsautoreport/slack.py [hunk 6]
```diff
@@ -42,62 +45,83 @@ def _upload_file(
...
-        response = self._client.files_upload(
-            channels=[self._channel_id],
-            file=file_path,
-            filename=filename,
-            initial_comment=file_title,
-            thread_ts=thread_time_stamp,
-        )
-        response.validate()
-        logger.info("%s was uploaded successfully", filename)
+        TRIES = 3
```
Comment (rccrdpccl) [inline src/jobsautoreport/slack.py:48]: Cant we use [retry package](https://pypi.org/project/retry/) instead?

Comment (osherdp) [reply src/jobsautoreport/slack.py:48]: +1

also, do we really need this kind of retry? I don't remember seeing any problem uploading files to slack in other places we do that

**Review by osherdp — CHANGES_REQUESTED** [review]

File: src/jobsautoreport/slack.py (hunk 6 above)
Comment (danmanor) [reply src/jobsautoreport/slack.py:48]: Lately we have been getting timeouts on upload tries to slack in stage environment from time time to time, see https://redhat-internal.slack.com/archives/C02L26KMHMW/p1681999420141539. I think if we can handle it, it would be better then letting the pod restart. WDYT?

Comment (osherdp) [reply src/jobsautoreport/slack.py:48]: yeah it's probably better to retry in that case
let's have it in another PR either way

**Review by osherdp — APPROVED** [review]
